src/gadgets/sumcheck.rs

```diff
@@ -0,0 +1,336 @@
+  let mut power = r.clone();
+  for (i, coeff) in coeffs.iter().enumerate().skip(1) {
+    // logic: eval_aux = eval + power * coeff
+    let eval_aux = eval.clone().add(&Num::<Scalar>::from(
+      power.mul(cs.namespace(|| format!("mul {}", i)), &coeff)?,
+    ));
+    // r1cs constr: power * coeff = eval - eval_aux
+    let allocated_eval: AllocatedNum<Scalar> =
+      AllocatedNum::<Scalar>::alloc(&mut cs, || Ok(eval.get_value().unwrap()))?;
+    let allocated_eval_aux: AllocatedNum<Scalar> =
+      AllocatedNum::<Scalar>::alloc(&mut cs, || Ok(eval_aux.get_value().unwrap()))?;
+    cs.enforce(
+      || format!("power * coeff = eval_aux - eval{}", i),
+      |lc| lc + power.get_variable(),
+      |lc| lc + coeff.get_variable(),
+      |lc| lc + allocated_eval_aux.get_variable() - allocated_eval.get_variable(),
+    );
+    eval = eval_aux;
+
+    // logic: power_aux = power * r
+    let power_aux = power.mul(cs.namespace(|| format!("mul2 {}", i)), &r)?;
+    // r1cs constr: power_aux = power * r
+    cs.enforce(
+      || format!("power_aux = power * r {}", i),
+      |lc| lc + power.get_variable(),
+      |lc| lc + r.get_variable(),
+      |lc| lc + power_aux.get_variable(),
+    );
+    power = power_aux;
+  }
+  let allocated_eval: AllocatedNum<Scalar> =
+    AllocatedNum::<Scalar>::alloc(&mut cs, || Ok(eval.get_value().unwrap()))?;
+
+  Ok(allocated_eval)
+}
+
+#[cfg(test)]
+mod tests {
+  use super::*;
+  use crate::bellperson::{
+    r1cs::{NovaShape, NovaWitness},
+    {shape_cs::ShapeCS, solver::SatisfyingAssignment},
+  };
+  use crate::spartan::sumcheck::UniPoly;
+  use pasta_curves::{arithmetic::CurveAffine, pallas, pallas::Scalar, vesta};
+
+  use crate::traits::{Group, TranscriptEngineTrait, TranscriptReprTrait};
+  type PastaG1 = pasta_curves::pallas::Point;
+  use crate::spartan::polynomial::MultilinearPolynomial;
+  use crate::spartan::sumcheck::SumcheckProof;
+
+  fn synthetize_uni_evaluate<Scalar, CS>(
+    mut cs: CS,
+    coeffs: Vec<Scalar>,
+    r: Scalar,
+  ) -> AllocatedNum<Scalar>
+  where
+    Scalar: PrimeField + PrimeFieldBits,
+    CS: ConstraintSystem<Scalar>,
+  {
+    // prepare inputs
+    let r_var = AllocatedNum::<Scalar>::alloc(cs.namespace(|| "alloc r"), || Ok(r)).unwrap();
+    let mut coeffs_var: Vec<AllocatedNum<Scalar>> = Vec::new();
+    for (i, coeff) in coeffs.iter().enumerate() {
+      coeffs_var.push(
+        AllocatedNum::<Scalar>::alloc(cs.namespace(|| format!("alloc coeffs {}", i)), || {
+          Ok(*coeff)
+        })
+        .unwrap(),
+      );
+    }
+
+    // define inputs
+    // let _ = r_var.inputize(cs.namespace(|| "Input r"));
+    // for (i, coeff) in coeffs_var.iter().enumerate() {
+    //   let _ = coeff.inputize(cs.namespace(|| format!("Input coeffs {}", i)));
+    // }
+
+    // evaluate in-circuit (synthetize)
+    let res = uni_evaluate(cs.namespace(|| "uni_evaluate"), coeffs_var, r_var).unwrap();
+    // let _ = res.inputize(cs.namespace(|| "Output res"));
```

Comment:
Leftover?

Reply:
Yes, the commented stuff will not be needed, leaving it for the moment meanwhile the PR is in draft mode, in order to check stuff while the PR is in progress.